### PR TITLE
rpc: avoid a panic in delayingConn

### LINKED ISSUE #63075: rpc: panic in delayingConn.Read (simulated latencies)
Found in #62435

```
=== RUN   TestTransientClusterSimulateLatencies/from_europe-west1
E210405 09:04:19.983639 10906 kv/kvserver/queue.go:1093  [n1,replicate,s1,r9/1:/Table/1{3-4}] 1  change replicas of r9 failed: fetching current range descriptor value: node unavailable; try another peer
panic: didn't get expected magic bytes header

goroutine 11367 [running]:
github.com/cockroachdb/cockroach/pkg/rpc.(*delayingConn).Read(0xc000fab440, 0xc001626fc0, 0x205, 0x205, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:913 +0x393
crypto/tls.(*atLeastReader).Read(0xc001037e60, 0xc001626fc0, 0x205, 0x205, 0xc001626fc0, 0x0, 0xc00e3346a8)
        /usr/local/go/src/crypto/tls/conn.go:779 +0x62
bytes.(*Buffer).ReadFrom(0xc009118600, 0x5398200, 0xc001037e60, 0x43c1a5, 0x40e0d60, 0x43bd6c0)
        /usr/local/go/src/bytes/buffer.go:204 +0xb1
crypto/tls.(*Conn).readFromUntil(0xc009118380, 0x7fcdce218250, 0xc000fab440, 0x5, 0xc000fab440, 0x4726dc)
        /usr/local/go/src/crypto/tls/conn.go:801 +0xf3
crypto/tls.(*Conn).readRecordOrCCS(0xc009118380, 0xc000c11300, 0xf4, 0x100)
        /usr/local/go/src/crypto/tls/conn.go:608 +0x115
crypto/tls.(*Conn).readRecord(...)
        /usr/local/go/src/crypto/tls/conn.go:576
```

The issue is that the code assumes no magic header == impossible and throws a panic in that case.

There are two problems here:

- even if the situation was indeed impossible, this code should return an assertion error, not throw a panic
- the situation is, in fact, possible, if the server shuts down prematurely while the rpc connection is being established.

### PR DESCRIPTION
**Pull request by knz**
Fixes #63075. 
Needed for #62435. 

If the remote server shuts down while the connection is established,
the delaying conn may not complete its handshake. Avoid a panic in
that case.

### CODE REVIEW

File: pkg/rpc/context.go [hunk 1]
```diff
@@ -910,7 +912,7 @@ func (d *delayingConn) Read(b []byte) (n int, err error) {
 		}
 		// If we somehow don't get our expected magic, throw an error.
 		if hdr.Magic != magic {
-			panic(errors.New("didn't get expected magic bytes header"))
+			return 0, errors.WithStack(errMagicNotFound)
```
Comment (otan) [inline pkg/rpc/context.go:915]: any reason not to use `errors.New("didn't get expected magic bytes header")` directly here?

**Review by otan — APPROVED** [review]
ah, i saw this but thought it was related to one of jordan's fixes. nice!

File: pkg/rpc/context.go (hunk 1 above)
Comment (knz) [reply pkg/rpc/context.go:915]: `errors.New` does more work

**Comment by knz** [comment]
>  i saw this but thought it was related to one of jordan's fixes

jordan fixed something else

**Comment by knz** [comment]
bors r=otan

**Comment by knz** [comment]
bors r=otan

**Pull request merged**